napari/layers/_shapes_layer/model.py

```diff
@@ -0,0 +1,1568 @@
+            Index of vertex if any that is at the coordinates. Returns `None`
+            if no vertex is found.
+        """
+        # Check selected shapes
+        if len(self.selected_shapes) > 0:
+            if self.mode == Mode.Select:
+                # Check if inside vertex of interaction box or rotation handle
+                inds = list(range(0, 8))
+                inds.append(9)
+                box = self._selected_box[inds]
+                distances = abs(box - coord[:2])
+
+                # Get the vertex sizes
+                rescale = self._get_rescale()
+                sizes = self._vertex_size*rescale/2
+
+                # Check if any matching vertices
+                matches = np.all(distances <= sizes, axis=1).nonzero()
+                if len(matches[0]) > 0:
+                    return self.selected_shapes[0], matches[0][-1]
+            elif self.mode in ([Mode.Direct, Mode.VertexInsert,
+                                Mode.VertexRemove]):
+                # Check if inside vertex of shape
+                inds = np.isin(self.data._index, self.selected_shapes)
+                vertices = self.data._vertices[inds]
+                distances = abs(vertices - coord[:2])
+
+                # Get the vertex sizes
+                rescale = self._get_rescale()
+                sizes = self._vertex_size*rescale/2
+
+                # Check if any matching vertices
+                matches = np.all(distances <= sizes, axis=1).nonzero()[0]
+                if len(matches) > 0:
+                    index = inds.nonzero()[0][matches[-1]]
+                    shape = self.data._index[index]
+                    _, idx = np.unique(self.data._index, return_index=True)
+                    return shape, index - idx[shape]
+
+        # Check if mouse inside shape
+        triangles = self.data._mesh_vertices[self.data._mesh_triangles]
+        inside = inside_triangles(triangles - coord[:2])
+        shapes = self.data._mesh_triangles_index[inside]
+
+        if len(shapes) > 0:
+            indices = shapes[:, 0]
+            z_list = self.data._z_order.tolist()
+            order_indices = np.array([z_list.index(m) for m in indices])
+            ordered_shapes = indices[np.argsort(order_indices)]
+            return ordered_shapes[0], None
+        else:
+            return None, None
+
+    def _shapes_in_box(self, corners):
+        """Determines which shapes, if any, are inside an axis aligned box
+
+        Parameters
+        ----------
+        corners : np.ndarray
+            2x2 array of two corners that will be used to create an axis
+            aligned box.
+
+        Returns
+        ----------
+        shapes : list
+            List of shapes that are inside the box.
+        """
+
+        triangles = self.data._mesh_vertices[self.data._mesh_triangles]
+        intersects = triangles_intersect_box(triangles, corners)
+        shapes = self.data._mesh_triangles_index[intersects, 0]
```

Comment:
Also I'd say that most of these calculating methods should be in a _shapes_layer_crontroller file. And the model file should only handle setting and getting properties on the shapes_layer

Reply:
I have moved some of this processing stuff to the `shape_list` class